File: Data-Structures-and-Algorithms/0151-ReverseWordsinaString/reverseWords.js
const reverseWords = function(s) {
    // split the string by space into an array O(N)
    let arr = s.split(' ');
    let res = '';
    // iterate backwards through the array O(N)
    for (var i = arr.length-1; i >= 0; i--) {
        // if the element is a word, add the word to the string 
        if (arr[i] != '') {
            // if is not the first word, add a space before the word to avoid trimming
            if (res.length > 0) {
                res += " "
            }
            res += arr[i]
        } 
    }
    return res
};

s = " the sky is blue "
console.log(reverseWords(s))

File: Data-Structures-and-Algorithms/0151-ReverseWordsinaString/reverseWords.py
class Solution(object):
    def reverseWords(self, s):
        # all three (split, reverse, and join) operations are O(N), so O(3N) ~= O(N)
        S = reversed(s.split())
        return " ".join(S)

s = " the sky is blue "
print(Solution().reverseWords(s))

# O(N)
# O(N) for space storing the array of s.split()
# Reverse words